src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -509,6 +509,16 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
 
         try:
             return_type = type_info.promote(left_type, right_type)
+
+            if isinstance(mask_type, ct.FieldType):
+                if isinstance(return_type, ct.ScalarType):
+                    tmp_type = return_type
+                elif isinstance(return_type, ct.FieldType):
+                    tmp_type = return_type.dtype
+                return_type = type_info.promote(
+                    return_type, ct.FieldType(dims=mask_type.dims, dtype=tmp_type)
+                )
+
```

Comment:
` if isinstance(return_type, ct.ScalarType) and isinstance(mask_type, ct.FieldType):
                
return_type = type_info.promote(
                    return_type, ct.FieldType(dims=mask_type.dims, dtype=return_type)
                )`

Reply:
done. thanks!

Reply:
@nfarabullini had to revert your suggested change, due to an additional test failing. See Hannes' comment below.